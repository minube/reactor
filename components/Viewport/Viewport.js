import { array, bool, func, node, number, object, oneOfType } from 'prop-types';
import React, { PureComponent } from 'react';
import { ScrollView, StyleSheet } from 'react-native';

import { LAYOUT } from '../../common';
import Button from '../Button';
import Motion from '../Motion';
import styles from './Viewport.style';

const MOTION_DURATION = 500;

class Viewport extends PureComponent {
  state = {
    scroll: 0,
  }

  _onScroll = ({ nativeEvent: { contentOffset: { y: scroll } } }) => {
    const { props: { onScroll } } = this;
    this.setState({ scroll });

    if (onScroll) onScroll(scroll);
  }

  render() {
    const {
      _onScroll,
      props: {
        children, onBack, style, styleContent, visible,
      },
      state: { scroll },
    } = this;

    const { VIEWPORT: { H: height, W: width } } = LAYOUT;

    return (
      <Motion
        disabled={false}
        style={StyleSheet.flatten([styles.container, { height, width }, style])}
        timeline={[{ property: 'translateX', value: visible ? 0 : width }]}
        type="timing"
      >
        <Motion
          delay={MOTION_DURATION / 2}
          style={StyleSheet.flatten([styles.content, styleContent])}
          timeline={[{ property: 'opacity', value: visible ? 1 : 0 }]}
        >
          <ScrollView onScroll={_onScroll}>
            {children}
          </ScrollView>
        </Motion>
        { onBack &&
          <Motion
            delay={MOTION_DURATION}
            timeline={[{ property: 'scale', value: (visible && scroll === 0) ? 1 : 0 }]}
            style={styles.buttonBack}
          >
            <Button icon="left" responsive onPress={onBack} />
          </Motion> }
      </Motion>
    );
  }
}

Viewport.propTypes = {
  children: node,
  onBack: func,
  onScroll: func,
  style: oneOfType([array, number, object]),
  styleContent: oneOfType([array, number, object]),
  visible: bool,
};

Viewport.defaultProps = {
  children: undefined,
  onBack: undefined,
  onScroll: undefined,
  style: [],
  styleContent: [],
  visible: true,
};

export default Viewport;
