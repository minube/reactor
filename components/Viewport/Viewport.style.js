import { StyleSheet } from 'react-native';

import { THEME } from '../../common';

const { COLOR, OFFSET } = THEME;

export default StyleSheet.create({
  container: {
    position: 'absolute',
    top: 0,
    left: 0,
    overflow: 'hidden',
    backgroundColor: COLOR.WHITE,
  },

  content: {
    position: 'absolute',
    height: '100%',
    width: '100%',
    padding: OFFSET,
    zIndex: 0,
  },

  buttonBack: {
    position: 'absolute',
    top: OFFSET,
    left: OFFSET,
    zIndex: 1,
  },
});
