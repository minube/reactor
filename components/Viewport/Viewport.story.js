import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { boolean, object } from '@storybook/addon-knobs/react';

import Viewport from './Viewport';
import { THEME } from '../../common';

const {
  BORDER_RADIUS, BUTTON, COLOR, FONT, UNIT,
} = THEME;

const STYLE = {
  width: 375,
  height: 667,
  backgroundColor: COLOR.CONTENT,
};

const STYLE_CONTENT = {
  padding: 32,
};

const STYLE_NAVIGATION = {
  overflow: 'hidden',
};

const CHILDREN =
  <Text>
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
    Why do we use it?
    It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout. The point of using Lorem Ipsum is that it has a more-or-less normal distribution of letters, as opposed to using 'Content here, content here', making it look like readable English. Many desktop publishing packages and web page editors now use Lorem Ipsum as their default model text, and a search for 'lorem ipsum' will uncover many web sites still in their infancy. Various versions have evolved over the years, sometimes by accident, sometimes on purpose (injected humour and the like).
    Where does it come from?
    Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words, consectetur, from a Lorem Ipsum passage, and going through the cites of the word in classical literature, discovered the undoubtable source. Lorem Ipsum comes from sections 1.10.32 and 1.10.33 of "de Finibus Bonorum et Malorum" (The Extremes of Good and Evil) by Cicero, written in 45 BC. This book is a treatise on the theory of ethics, very popular during the Renaissance. The first line of Lorem Ipsum, "Lorem ipsum dolor sit amet..", comes from a line in section 1.10.32.
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
    Why do we use it?
    It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout. The point of using Lorem Ipsum is that it has a more-or-less normal distribution of letters, as opposed to using 'Content here, content here', making it look like readable English. Many desktop publishing packages and web page editors now use Lorem Ipsum as their default model text, and a search for 'lorem ipsum' will uncover many web sites still in their infancy. Various versions have evolved over the years, sometimes by accident, sometimes on purpose (injected humour and the like).
    Where does it come from?
    Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words, consectetur, from a Lorem Ipsum passage, and going through the cites of the word in classical literature, discovered the undoubtable source. Lorem Ipsum comes from sections 1.10.32 and 1.10.33 of "de Finibus Bonorum et Malorum" (The Extremes of Good and Evil) by Cicero, written in 45 BC. This book is a treatise on the theory of ethics, very popular during the Renaissance. The first line of Lorem Ipsum, "Lorem ipsum dolor sit amet..", comes from a line in section 1.10.32.
  </Text>;

storiesOf('🚀 Viewport', module)
  .addWithJSX('default', () => (
    <Viewport>{CHILDREN}</Viewport>
  ))
  .addWithJSX('scroll (false)', () => (
    <Viewport scroll={false}>{CHILDREN}</Viewport>
  ))
  .addWithJSX('⚡ onPress', () => (
    <Viewport onBack={action('Viewport.onBack()')}>{CHILDREN}</Viewport>
  ))
  .addWithJSX('style', () => (
    <Viewport style={STYLE} styleContent={STYLE_CONTENT}>{CHILDREN}</Viewport>
  ))
  .addWithJSX('🏀 Playground', () => (
    <View style={StyleSheet.flatten([STYLE, STYLE_NAVIGATION])}>
      <Viewport style={STYLE} visible>
        <Text>viewport.1</Text>
      </Viewport>
      <Viewport
        onBack={action('Viewport.onBack()')}
        onScroll={action('Viewport.onScroll()')}
        scroll={boolean('scroll', true)}
        style={STYLE}
        styleContent={object('style', STYLE_CONTENT, 'viewport-style')}
        visible={boolean('visible', true)}
      >
        {CHILDREN}
      </Viewport>
    </View>
  ))


