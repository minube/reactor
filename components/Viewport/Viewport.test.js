import React from 'react';
import { Text } from 'react-native';
import TestRenderer from 'react-test-renderer';

import Viewport from './Viewport';

const VIDEO_PLACEHOLDER = 'http://download.blender.org/peach/bigbuckbunny_movies/BigBuckBunny_320x180.mp4';

describe('<Viewport>', () => {
  it('renders', () => {
    const renderer = TestRenderer.create(<Viewport />).toJSON();
    expect(renderer).toMatchSnapshot();
  });

  it('when {children}', () => {
    const renderer = TestRenderer.create(<Viewport><Text>Hello World</Text></Viewport>).toJSON();
    expect(renderer).toMatchSnapshot();
  });

  it('when {onBack}', () => {
    const renderer = TestRenderer.create(<Viewport onBack={() => {}} />).toJSON();
    expect(renderer).toMatchSnapshot();
  });

  it('when {onScroll}', () => {
    const renderer = TestRenderer.create(<Viewport onScroll={() => {}} />).toJSON();
    expect(renderer).toMatchSnapshot();
  });

  it('when {style}', () => {
    const renderer = TestRenderer.create(<Viewport style={{ width: 320, height: 480 }} />).toJSON();
    expect(renderer).toMatchSnapshot();
  });

  it('when {styleContent}', () => {
    const renderer = TestRenderer.create(<Viewport styleContent={{ padding: 32 }} />).toJSON();
    expect(renderer).toMatchSnapshot();
  });

  it('when {visible}', () => {
    const renderer = TestRenderer.create(<Viewport visible />).toJSON();
    expect(renderer).toMatchSnapshot();
  });
});
