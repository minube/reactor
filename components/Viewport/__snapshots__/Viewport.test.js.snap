// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Viewport> renders 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "height": 1334,
      "left": 0,
      "overflow": "hidden",
      "position": "absolute",
      "top": 0,
      "transform": Array [
        Object {
          "translateX": 0,
        },
      ],
      "width": 750,
    }
  }
>
  <View
    collapsable={undefined}
    style={
      Object {
        "height": "100%",
        "opacity": 1,
        "padding": 16,
        "position": "absolute",
        "width": "100%",
        "zIndex": 0,
      }
    }
  >
    <RCTScrollView
      onScroll={[Function]}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;

exports[`<Viewport> when {children} 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "height": 1334,
      "left": 0,
      "overflow": "hidden",
      "position": "absolute",
      "top": 0,
      "transform": Array [
        Object {
          "translateX": 0,
        },
      ],
      "width": 750,
    }
  }
>
  <View
    collapsable={undefined}
    style={
      Object {
        "height": "100%",
        "opacity": 1,
        "padding": 16,
        "position": "absolute",
        "width": "100%",
        "zIndex": 0,
      }
    }
  >
    <RCTScrollView
      onScroll={[Function]}
    >
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          Hello World
        </Text>
      </View>
    </RCTScrollView>
  </View>
</View>
`;

exports[`<Viewport> when {onBack} 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "height": 1334,
      "left": 0,
      "overflow": "hidden",
      "position": "absolute",
      "top": 0,
      "transform": Array [
        Object {
          "translateX": 0,
        },
      ],
      "width": 750,
    }
  }
>
  <View
    collapsable={undefined}
    style={
      Object {
        "height": "100%",
        "opacity": 1,
        "padding": 16,
        "position": "absolute",
        "width": "100%",
        "zIndex": 0,
      }
    }
  >
    <RCTScrollView
      onScroll={[Function]}
    >
      <View />
    </RCTScrollView>
  </View>
  <View
    collapsable={undefined}
    style={
      Object {
        "left": 16,
        "position": "absolute",
        "top": 16,
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
        "zIndex": 1,
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderRadius": 30,
          "height": 60,
          "margin": 5,
          "paddingHorizontal": 0,
          "width": 60,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Object {
            "alignContent": "center",
            "alignItems": "center",
            "backgroundColor": "#4a4a4a",
            "borderRadius": 30,
            "display": "flex",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "center",
            "paddingHorizontal": 0,
            "width": 60,
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "testUri": "../../../components/Icon/assets/left.png",
            }
          }
          style={
            Object {
              "backgroundColor": "transparent",
              "height": 20,
              "width": 20,
            }
          }
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`<Viewport> when {onScroll} 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "height": 1334,
      "left": 0,
      "overflow": "hidden",
      "position": "absolute",
      "top": 0,
      "transform": Array [
        Object {
          "translateX": 0,
        },
      ],
      "width": 750,
    }
  }
>
  <View
    collapsable={undefined}
    style={
      Object {
        "height": "100%",
        "opacity": 1,
        "padding": 16,
        "position": "absolute",
        "width": "100%",
        "zIndex": 0,
      }
    }
  >
    <RCTScrollView
      onScroll={[Function]}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;

exports[`<Viewport> when {style} 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "height": 480,
      "left": 0,
      "overflow": "hidden",
      "position": "absolute",
      "top": 0,
      "transform": Array [
        Object {
          "translateX": 0,
        },
      ],
      "width": 320,
    }
  }
>
  <View
    collapsable={undefined}
    style={
      Object {
        "height": "100%",
        "opacity": 1,
        "padding": 16,
        "position": "absolute",
        "width": "100%",
        "zIndex": 0,
      }
    }
  >
    <RCTScrollView
      onScroll={[Function]}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;

exports[`<Viewport> when {styleContent} 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "height": 1334,
      "left": 0,
      "overflow": "hidden",
      "position": "absolute",
      "top": 0,
      "transform": Array [
        Object {
          "translateX": 0,
        },
      ],
      "width": 750,
    }
  }
>
  <View
    collapsable={undefined}
    style={
      Object {
        "height": "100%",
        "opacity": 1,
        "padding": 32,
        "position": "absolute",
        "width": "100%",
        "zIndex": 0,
      }
    }
  >
    <RCTScrollView
      onScroll={[Function]}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;

exports[`<Viewport> when {visible} 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "height": 1334,
      "left": 0,
      "overflow": "hidden",
      "position": "absolute",
      "top": 0,
      "transform": Array [
        Object {
          "translateX": 0,
        },
      ],
      "width": 750,
    }
  }
>
  <View
    collapsable={undefined}
    style={
      Object {
        "height": "100%",
        "opacity": 1,
        "padding": 16,
        "position": "absolute",
        "width": "100%",
        "zIndex": 0,
      }
    }
  >
    <RCTScrollView
      onScroll={[Function]}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;
